tests/instruments/test_user_api.py

```diff
@@ -0,0 +1,125 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import pytest
+import serial
+
+from pymeasure.adapters import SerialAdapter, VISAAdapter
+from pymeasure.instruments import Instrument
+from pyvisa.constants import InterfaceType
+
+
+class MultiprotocolInstrument(Instrument):
+    """Test instrument for testing the configuration and use of multiple connection methods.
+
+    The instrument support multiple transports methods(serial, TCP/IP, GPIB).
+    In case of serial connections, the default baud rate is low (2400), but configurable in
+    the device, so it can use faster communication if desired.
+    Two examples in our instruments are AMI430, Keithley2260B (for TCP/IP, there are more
+    with GPIB capability)
+
+    It also has an attribute to configure a non-serial (gpib) connection. This demonstrates
+    that there may be attributes that are not valid for a serial connection, only others.
+    """
+    def __init__(self, adapter, name="Instrument with multiple connection methods", **kwargs):
+        # As an instrument contributor, I want to define default connection settings for my
+        # instrument with a minimum of boiler-plate. These settings should be easily
+        # user-overridable if appropriate without having to change the code/implementation
+        # of my instrument.
+
+        super().__init__(adapter,
+                         name=name,
+                         **kwargs)
+        if isinstance(adapter, (int, str)):
+            # These are only cases where adapter is created by Instrument class, in this case
+            # the adapter is a VISAAdapter class based on pyvisa
+            if_type = self.adapter.connection.resource_info.interface_type
+            # Different communication methods might have different valid and/or needed attributes
+            # This is just an example on how this is possible since pyvisa includes
+            # several type of interfaces
+            if if_type is InterfaceType.asrl and not 'baud_rate' in kwargs:
+                self.adapter.connection.baud_rate = 2400
+            elif if_type is InterfaceType.gpib and not 'enable_repeat_addressing' in kwargs:
+                self.adapter.connection.enable_repeat_addressing = False
```

Comment:
OK, that is an interesting approach I hadn't considered. I have to say, though, that we should not have this much boilerplate in instrument implementations that need this (especially compared to the 2 lines I had before) - this is hard to get right and keep consistent.

By now I'm confident that we can allow instrument implementations to define ~any arguments to pass to `super().__init__`, and we can push this handling down into `VISAAdapter`, ~where you deleted the safeKeywords stuff.

I have sketched this approach in the afternoon, handles both kwarg user typos and finding the correct set of args for a given resource:
```
        self.manager = pyvisa.ResourceManager(visa_library)

        def valid_visa_kwargs(resource_manager):
            """Return all possible attribute names of registered resource classes."""
            # TODO move function outside when finished
            valid = set()
            for resource_class in resource_manager._resource_classes.values():
                for attr in resource_class:  #TODO: enumerate attrs and properties
                    # maybe need to filter attributes and properties
                    valid += v
            return valid

        invalid_kwargs = set(kwargs.keys()) - valid_visa_kwargs(self.manager)
        if invalid_kwargs:
            raise ValueError(f"Invalid argument(s) passed to VISAAdapter: {invalid_kwargs}")

        while True:
            try:
                self.connection = self.manager.open_resource(
                    resource_name,
                    **kwargs
                )
                break  # happy path, we suceeded
            # WIP: if any kwargs are inappropriate for the present resource, pop off and retry; we get this from pyvisa:
            # ValueError(
            #     "%r is not a valid attribute for type %s"
            #     % (key, res.__class__.__name__)
            # )
            except ValueError as exc:
                false_arg = ''  # TODO: parse arg name from exc message
                kwargs.pop(false_arg)
```

Maybe some aspects of your approach can be used to make this smaller or so, but I already believe that just retrying is simplest (this is Python, after all).

If you want you can take this and finish it, or I do it later (tomorrow), to include in this branch.

Reply:
Everything boils down to allow instrument contributors to easily and clearly define appropriate defaults for different VISAAdapter interface in the same instrument.
I have a different proposal, based on the concept that VISAAdapter cannot properly fiddle with parameters, but just obey to them.
I have added a static method to VISAAdapter class to help setting the defaults.
The final result is as follow:
```
        VISAAdapter.setdefaults(adapter, InterfaceType.asrl, kwargs,
                                baud_rate = 2400)
        VISAAdapter.setdefaults(adapter, InterfaceType.gpib, kwargs,
                                enable_repeat_addressing = False)
```

Please have a look and let me know.

Reply:
Sure, I'll take a look!